include/connection.h

```diff
@@ -50,10 +49,27 @@ typedef struct NodeConnectionEntry
 } NodeConnectionEntry;
 
 
+
+typedef struct {
+	ShardId shardId;
+	int nReplicas;
+	bool* status;
+	PGconn** conn;
+} ShardConnections;
+
 /* function declarations for obtaining and using a connection */
 extern PGconn * GetConnection(char *nodeName, int32 nodePort);
 extern void PurgeConnection(PGconn *connection);
 extern void ReportRemoteError(PGconn *connection, PGresult *result);
+extern PGconn* ConnectToNode(char *nodeName, char *nodePort);
+
+typedef bool (*ShardAction)(ShardId id, PGconn* conn, void* arg, bool status);
 
+/* 
+ * Perform action for all shards and all shard replicas.
+ * Returns number of shard for which operation od failed or INVALID_SHARD_ID in case of success
+ */
+struct HTAB;
+extern ShardId DoForAllShards(struct HTAB* shardConnHash, ShardAction action, void* arg);
```

Comment:
I'd prefer not to require all callers to construct an HTAB, could this be changed to a List *shardConnList? 

(the COPY implementation will have the convert the hash table to a list, but this is a very minor overhead)